### PR TITLE
fix for #5503

### PR DESCRIPTION
**Pull request by icesmurf**
First cut at a patch for ruby connection timeouts when establishing a tcp connection, seems to work as expected here.. I've tested for most of the stuff i was having issues with, and it doesn't seem to have broken anything.. not sure about writing spec tests or anything?

### CODE REVIEW

**Comment by danielmewes** [comment]
Nice, thanks for the PR @icesmurf .

If you haven't done this yet, would you mind going to http://rethinkdb.com/community/cla/ and signing the CLA? We require this before we can merge any PR.

Pinging @mlucy for reviewing your patch.

**Comment by icesmurf** [comment]
Yeah, i've already signed it. :-)

File: drivers/ruby/lib/net.rb [hunk 1]
```diff
@@ -677,20 +677,40 @@ def init_socket
...
         verify_cert!(@socket, context)
       else
         @socket = base_socket
+        socket_connect
       end
     end
 
     def base_socket
-      socket = TCPSocket.open(@host, @port)
+      @socket.close if @socket
```
Comment (mlucy) [inline drivers/ruby/lib/net.rb]: I don't think we should be calling `@socket.close` here.  This was previously a pure function, so it's a little weird to have it modifying a member variable.  If we want to be closing the socket, I think we should be doing it in `init_socket`.  In this case, though, `init_socket` is only called by `connect`, which checks that `@socket` isn't set immediately before calling it on line 663, so we should probably just omit this conditional close entirely.

File: drivers/ruby/lib/net.rb [hunk 2]
```diff
@@ -677,20 +677,40 @@ def init_socket
         context = create_context(@ssl_opts)
         @socket = OpenSSL::SSL::SSLSocket.new(@tcp_socket, context)
         @socket.sync_close = true
-        @socket.connect
+        socket_connect
```
Comment (mlucy) [inline drivers/ruby/lib/net.rb]: I'm a little confused by this logic.  Below we call `socket_connect` after `base_socket`, but here we call `base_socket`, then put that into an `SSLSocket`, then call `socket_connect` on that.  Before we followed the same code path here for `@tcp_socket` as we did for `@socket` below, why did that change?

(Also, just to double-check, did you test these changes with both SSL and non-SSL connections?)

File: drivers/ruby/lib/net.rb [hunk 3]
```diff
@@ -677,20 +677,40 @@ def init_socket
...
     def base_socket
-      socket = TCPSocket.open(@host, @port)
+      @socket.close if @socket
+      socket = Socket.new(Socket::AF_INET, Socket::SOCK_STREAM, 0)
       socket.setsockopt(Socket::IPPROTO_TCP, Socket::TCP_NODELAY, 1)
       socket.setsockopt(Socket::SOL_SOCKET, Socket::SO_KEEPALIVE, 1)
       socket
     end
 
+    def socket_connect
```
Comment (mlucy) [inline drivers/ruby/lib/net.rb]: Could you add a comment explaining what this code is doing and why it solves the timeout issue?

File: drivers/ruby/lib/net.rb [hunk 4]
```diff
@@ -677,20 +677,40 @@ def init_socket
...
+      address = Socket.getaddrinfo(@host, nil, Socket::AF_INET).first[3]
+      sockaddr = Socket.pack_sockaddr_in(@port, address)
+      begin
+        @socket.connect_nonblock(sockaddr)
+      rescue Errno::EINPROGRESS
+        IO.select([@socket], [@socket], [@socket], @timeout)
+        begin
+          @socket.connect_nonblock(sockaddr)
+        rescue Errno::EISCONN
+          return
```
Comment (mlucy) [inline drivers/ruby/lib/net.rb]: This is the case where we've failed to connect, right?  Shouldn't we be throwing an exception?

File: drivers/ruby/lib/net.rb [hunk 5]
```diff
@@ -677,20 +677,40 @@ def init_socket
...
+        rescue Errno::EALREADY
+          @socket.close
+          raise ReqlRuntimeError, "Connection timed out"
```
Comment (mlucy) [inline drivers/ruby/lib/net.rb]: Previously we were raising `Timeout::Error` for connection timeouts.  We should probably continue to do that.

File: drivers/ruby/lib/net.rb [hunk 6]
```diff
@@ -677,20 +677,40 @@ def init_socket
...
       socket
     end
 
+    def socket_connect 
+      address = Socket.getaddrinfo(@host, nil, Socket::AF_INET).first[3]
+      sockaddr = Socket.pack_sockaddr_in(@port, address)
+      begin
+        @socket.connect_nonblock(sockaddr)
+      rescue Errno::EINPROGRESS
+        IO.select([@socket], [@socket], [@socket], @timeout)
```
Comment (mlucy) [inline drivers/ruby/lib/net.rb]: The documentation for `select` says that it will return `nil` if the timeout value is given and no object is ready in that number of seconds.  It seems clearer to check that and raise the timeout error here.

File: drivers/ruby/lib/net.rb [hunk 7]
```diff
@@ -677,20 +677,40 @@ def init_socket
...
+        begin
+          @socket.connect_nonblock(sockaddr)
+        rescue Errno::EISCONN
+          return
+        rescue Errno::EALREADY
```
Comment (mlucy) [inline drivers/ruby/lib/net.rb]: Why does `Errno::EALREADY` indicate a timeout?  The documentation seems to say that `EALREADY` is the same as `EINVAL`, which means "the `socket` is a listening socket".

File: drivers/ruby/lib/net.rb [hunk 8]
```diff
@@ -677,20 +677,40 @@ def init_socket
...
       socket.setsockopt(Socket::IPPROTO_TCP, Socket::TCP_NODELAY, 1)
       socket.setsockopt(Socket::SOL_SOCKET, Socket::SO_KEEPALIVE, 1)
       socket
     end
 
+    def socket_connect 
+      address = Socket.getaddrinfo(@host, nil, Socket::AF_INET).first[3]
+      sockaddr = Socket.pack_sockaddr_in(@port, address)
+      begin
+        @socket.connect_nonblock(sockaddr)
```
Comment (mlucy) [inline drivers/ruby/lib/net.rb]: Why do we call `connect_nonblock`, and if we get `EINPROGRESS` call `select` and then `connect_nonblock` again?  Couldn't we just call `select` first, followed by `connect_nonblock` once we know it's safe?

File: drivers/ruby/lib/net.rb [hunk 9]
```diff
@@ -677,20 +677,40 @@ def init_socket
...
+      @socket.close if @socket
+      socket = Socket.new(Socket::AF_INET, Socket::SOCK_STREAM, 0)
       socket.setsockopt(Socket::IPPROTO_TCP, Socket::TCP_NODELAY, 1)
       socket.setsockopt(Socket::SOL_SOCKET, Socket::SO_KEEPALIVE, 1)
       socket
     end
 
+    def socket_connect 
+      address = Socket.getaddrinfo(@host, nil, Socket::AF_INET).first[3]
+      sockaddr = Socket.pack_sockaddr_in(@port, address)
```
Comment (mlucy) [inline drivers/ruby/lib/net.rb]: Why is this necessary?  Does `Socket.sockaddr_in(@port, @host)` not work for some reason?

File: drivers/ruby/lib/net.rb (hunk 3 above)
Comment (mlucy) [inline drivers/ruby/lib/net.rb]: What's the reason why this is called by `init_socket` rather than `base_socket`?  Is it not correct to finish connecting the socket before `base_socket` returns?

**Comment by mlucy** [comment]
Thanks for the pull request @icesmurf !  There are a few things I was unclear on in it.  Sorry for all the comments, but I'm a little paranoid about connection logic like this, and I'd like to leave as little magic in the code as possible.

File: drivers/ruby/lib/net.rb (hunk 1 above)
Comment (icesmurf) [inline drivers/ruby/lib/net.rb]: agree, i'd fixed this already, not sure why it wasn't in the PR, must have messed up my squash

File: drivers/ruby/lib/net.rb (hunk 3 above)
Comment (icesmurf) [inline drivers/ruby/lib/net.rb]: for some reason i thought it was better to do it individually, but i don't think it matters.

File: drivers/ruby/lib/net.rb (hunk 5 above)
Comment (icesmurf) [inline drivers/ruby/lib/net.rb]: agree

File: drivers/ruby/lib/net.rb (hunk 7 above)
Comment (icesmurf) [inline drivers/ruby/lib/net.rb]: heh, been trying to get my head around that as well actually, the examples i was looking at were terrible.

**Comment by icesmurf** [comment]
Agree with mlucy on alot of his comments, i've thrown most of it out ahem :-) 

I think this method is ALOT simpler and more understandable at a glance, referenced from ruby "Socket" documentation, not sure if we need another connect_nonblock and rescue after, it seems redundant to me, but they do that in the Socket examples, thoughts?

**Comment by danielmewes** [comment]
@mlucy Could you take another look at this please?

**Comment by icesmurf** [comment]
Actually, hold off, i've noticed a few issues i'm looking at, esp in
consideration of next, i've had some issues using non-blocking sockets with
mass updates and am trying to figure out why.

**Comment by srh** [comment]
Since this is old and the author says it has issues, closing.
